Track best-possible-headers (TheBlueMatt)

Sjors opened the pull request:
Rebase of #12138:

> This adds a `setBlockIndexHeaderCandidates` which mimics `setBlockIndexCandidates` and is
The set of all leaf `CBlockIndex` entries with` BLOCK_VALID_TREE` (for itself and all ancestors) and
as good as our current tip or better. Entries here are potential future candidates for insertion
into `setBlockIndexCandidates`, once we get all the required block data. Thus, entries here
represent chains on which we should be actively downloading block data.
>
> Note that we define "as good as our current tip or better" slightly differently here than in
setBlockIndexCandidates - we include things which will have a higher nSequence (but have the
same chain work) here, but do not include such entries in setBlockIndexCandidates. This is
because we prefer to also download towards chains which have the same total work as our current
chain (as an optimization since a reorg is very possible in such cases).
>
> Note that, unlike `setBlockIndexCandidates`, we only store "leaf" entries here, as we are not as
aggressively prune-able (`setBlockIndexCandidates` are things which we can, and usually do, try to connect immediately, and thus entries dont stick around for long). Thus, it may be the case that
`chainActive.Tip()` is NOT in `setBlockIndexHeaderCandidates`.
>
> Additionally, unlike setBlockIndexCandidates, we are happy to store entries which are not
connectable due to pruning here.
>
> This is useful as it (finally) disconnects net_processing logic from the "store on disk" logic in validation.cpp. More importantly, it represents what you'd need from the consensus logic to implement a spv-first sync mode, as this provides a best-header which will follow invalidity - always pointing to the best-possible header even after block(s) are found to be invalid.

Not included:
* #13912 (test framework change)
* #13930 (source code comment which is useful on its own)
* a9db3dada0119c183d16627805e90c4dbca05c6a "Do not unlock cs_main in ABC unless we've actually made progress": already merged as part of #13023

Noteble rebase changes:
* `pindex_was_in_chain` is still needed due to  #12431, in order to notify the gui about a new block tip, only if the active chain was modified
* `qt/clientmodel.cpp` changed significantly, because things were moved to `interfaces/node.cpp`
* updated `ProcessNewBlock` bench

I added an additional commit, because the new `!fHasMoreOrSameWork && !parent_of_header_candidate` check in `AcceptBlock` can cause `submitblock` (modified in #13439) to mistakingly believe the block is a duplicate.

Added debug log statements to `rpc_preciousblock.py` that helped me figure out the above issue.

TheBlueMatt committed:
Require setBlockIndexCandidates be !BLOCK_FAILED_MASK

When we find an invalid block, instead of adding BLOCK_FAILED_CHILD
to its descendants in FindMostWorkChain, iterate
setBlockIndexCandidates to find candidate descendants and mark them
as BLOCK_FAILED_CHILD immediately, removing them from
setBlockIndexCandidates as we go. This keeps BLOCK_FAILED_MASK
entries out of setBlockIndexCandidates.

This also adds a few checks to CheckBlockIndex, including one which
checks that blocks with BLOCK_FAILED_CHILD are not, themselves,
marked invalid, but have an invalid parent. This should be fine for
most block indexes, however InvalidateBlock previously violated
this. Luckily most users shouldn't be running with -checkblockindex

Note that this introduces a bug where a block who's header was
received but data was not when a ancestor was found to be invalid
will not be marked BLOCK_FAILED_CHILD. Thus, when that block is
received, it will be added to setBlockIndexCandidates, violating
the new invariant. This is fixed in the next commit.

TheBlueMatt committed:
Add a set to track potential future chain tips based only on SPV.

This mirrors setBlockIndexCandidates but for BLOCK_VALID_TREE
instead of BLOCK_VALID_TRANSACTIONS && nTx. There are a few
differences between the two to keep the new set practical, see
code comments for more.

TheBlueMatt committed:
Stop using pindexBestHeader outside of validation.cpp

pindexBestHeader never considers the invalidity of a chain, only
the validity of the header chain. Using it for sync estimation and
GETHEADERS etc requests in net makes no sense. Instead, use the
new setBlockIndexHeaderCandidates to find the best candidate tip.

TheBlueMatt committed:
Replace ProcessNewBlock's fRequested with candidate header check

Instead of allowing net_processing to inform AcceptBlock as to
whether a block is interesting enough to commit to disk, use the
new setBlockIndexHeadersCandidates to determine if they lead
towards a chain with at least the same (or more) work as our
current tip. This further decouples the maze in net_processing from
our consensus anti-DoS logic.

Sjors committed:
[test] precious_block: additional debug log statements

Sjors commented:
cc @TheBlueMatt, @MarcoFalke, @jamesob as this touches your work

cc @jonasschnelli  since you reviewed the original PR (@skeees did too, but only the commit that I dropped)

Sjors commented on src/validation.cpp:
```diff
@@ -1290,6 +1319,130 @@ static void CheckForkWarningConditionsOnNewFork(CBlockIndex* pindexNewForkTip)
+            forward_it--; // Now points to it
+            forward_it = set_candidates.erase(forward_it);
+            it = std::set<CBlockIndex*, CBlockIndexWorkComparator>::reverse_iterator(forward_it);
+            // forward_it == it.base() now points to one-past previous it, making it point to one-before previous it.
```
Note: I find the above reverse-to-forward iterator dance remarkably confusing.

donaloconnor commented on src/validation.cpp:
```diff
@@ -1290,6 +1319,130 @@ static void CheckForkWarningConditionsOnNewFork(CBlockIndex* pindexNewForkTip)
+    LOCK(cs_main);
+    auto it = setBlockIndexHeaderCandidates.rbegin();
+    if (it == setBlockIndexHeaderCandidates.rend())
+        return nullptr;
```
Numerous places call this function but no nullptr check is performed and dereferences are executed on the ptr.
Can this return null in any case where this function is called?

Sjors replied:
That seems problematic indeed. There's an `assert` for this in https://github.com/bitcoin/bitcoin/blame/7b1ab22384d944f638c514f838e25b391204ef67/src/net_processing.cpp#L3308, but not in the calls from validation.cpp:
* in `NotifyHeaderTip()`: here it's compared with `pindexHeaderOld` which is a `nullptr` initially, but would non-null the next time. It's not dereferenced in that function, but it is passed on to `uiInterface.NotifyHeaderTip`. I could add an assert inside`if (fNotify)`
* in `IsCurrentForFeeEstimation()`: I could add an `assert` directly above the if statement that deferences it

The comment in `InvalidBlockFound` suggests that sometimes `setBlockIndexHeaderCandidates` can be empty, so the above wouldn't be enough.

donaloconnor commented on src/interfaces/node.cpp:
```diff
@@ -152,10 +152,10 @@ class NodeImpl : public Node
-            height = ::pindexBestHeader->nHeight;
-            block_time = ::pindexBestHeader->GetBlockTime();
+        const CBlockIndex* best_header = ::GetBestHeader();
+        if (best_header) {
```
Possible race condition here since we don't lock cs_main? Can the best_head data be deleted between time we call GetBestHeader() and when we use the ptr?

I'm not familiar with the code enough to know the answer to this so just putting it here in case.

Sjors committed:
[validation] explictly handle GetBestHeader() == nullptr

Sjors committed:
[rpc] explictly handle GetBestHeader() == nullptr

Sjors commented:
Rebased.

I added commits to explicitly handle cases where `GetBestHeader()` returns `nullptr` in `validation.cpp`. Both cases are UI related and don't seem very important, so they're handled in a fairly lazy way.

I added a commit to make header count fall back to block height in `getblockchaininfo` (`rpc/blockchain.cpp`).

That leaves `net_processing.cpp`, where I'm not sure what to do:

* one existing assert in:
 https://github.com/bitcoin/bitcoin/blob/9ce44d8156576c58390a10007b60f507d107d042/src/net_processing.cpp#L3308
* one deference `->GetBlockTime()` in `BlockRequestAllowed`
* one dereference in `->GetBlockTime()`` in `ProcessGetBlockData`
* one passing the pointer on to `connman` in `ProcessHeadersMessage`
* passing into `chainActive.GetLocator` in `ProcessMessage`

I'm not sure if those are safe.

Sjors commented:
Rebased. I'll hold back on addressing specific feedback until there's some high level agreement on whether this is useful at all.

practicalswift commented on src/validation.cpp:
```diff
@@ -1281,6 +1315,130 @@ static void CheckForkWarningConditionsOnNewFork(CBlockIndex* pindexNewForkTip) E
+                // to do here.
+                setBlockIndexHeaderCandidates.erase(it);
+                break;
+                parent_present = true;
```
This is dead code. The `parent_present = true;` is not meant to be after `break;` in this scope? :-)

practicalswift commented on src/validation.cpp:
```diff
@@ -1281,6 +1315,130 @@ static void CheckForkWarningConditionsOnNewFork(CBlockIndex* pindexNewForkTip) E
+            if ((*it)->GetAncestor(pindex->nHeight) == pindex) {
+                // pindex is useless - even if there are other tips based on it
+                // which we want in setBlockIndexHeaderCandidates, we're not
+                // gonna find them here.
```
"going to find them here." :-)

practicalswift commented on src/validation.cpp:
```diff
@@ -1281,6 +1315,130 @@ static void CheckForkWarningConditionsOnNewFork(CBlockIndex* pindexNewForkTip) E
+ *
+ * Works even if chainActive is empty!
+ *
+ * If chain_ordered_inertion, we assume that if pindex->pprev was previously a
```
Should be insertion?

practicalswift commented on src/validation.cpp:
```diff
@@ -107,10 +108,36 @@ class CChainState {
+     *
+     * Note that, unlike setBlockIndexCandidates, we only store "leaf" entries here, as we are not as
+     * aggressively prune-able (setBlockIndexCandidates are things which we can, and usually do, try to
+     * connect immediately, and thus entries dont stick around for long). Thus, it may be the case that
```
"Don't" :-)

practicalswift commented on src/validation.cpp:
```diff
@@ -3470,31 +3675,38 @@ bool CChainState::AcceptBlock(const std::shared_ptr<const CBlock>& pblock, CVali
+            }
+        }
+    }
+    if (!fHasMoreOrSameWork && !parent_of_header_candidate) return true; // We dont think this block leads somewhere interesting
```
"Don't" :-)

practicalswift commented on src/validation.cpp:
```diff
@@ -1281,6 +1315,130 @@ static void CheckForkWarningConditionsOnNewFork(CBlockIndex* pindexNewForkTip) E
+    if (lower_work) return; // We don't want things with less work than our current tip
+
+    bool parent_present = false;
+    if (pindex->pprev && setBlockIndexHeaderCandidates.count(pindex->pprev)) {
```
Nit: `setBlockIndexHeaderCandidates.count(pindex->pprev) > 0` to improve readability.

practicalswift commented on src/validation.cpp:
```diff
@@ -2958,10 +3155,18 @@ void CChainState::ReceivedBlockTransactions(const CBlock& block, CBlockIndex* pi
             queue.pop_front();
             pindex->nChainTx = (pindex->pprev ? pindex->pprev->nChainTx : 0) + pindex->nTx;
+            // Make sure to remove from sets which are indexed by nSequenceId first...
+            bool was_header_candidate = setBlockIndexHeaderCandidates.erase(pindex);
```
`setBlockIndexHeaderCandidates.erase(pindex) > 0` to improve readability and avoid implicit conversion.

practicalswift commented on src/validation.cpp:
```diff
@@ -4550,22 +4771,46 @@ void CChainState::CheckBlockIndex(const Consensus::Params& consensusParams)
+            // than chainActive.Tip(), it must be in setBlockIndexHeaderCandidates.
+            must_be_header_candidate_if_leaf = true;
+        }
+        if (setBlockIndexHeaderCandidates.count(pindex)) {
```
`setBlockIndexHeaderCandidates.count(pindex) > 0` to improve readability and avoid implicit conversion :-)

Sjors commented:
Abandoning for now. Consider it up for grabs, but I suggest getting concept ACKs before trying to rebase.

Sjors commented:
@ryanofsky it started out here: https://github.com/bitcoin/bitcoin/pull/9483#issuecomment-349790765
